internal/pkg/configloader/config.go

```diff
@@ -20,195 +19,191 @@ var (
+		toolMap[tool.Key()] = true
 	}
-	// if var file is empty, just return ""
-	if gConfig.VarFile != "" {
-		absVarFilePath, err = parseCustomPath(configFileName, gConfig.VarFile)
-		if err != nil {
-			return "", "", err
+
+	for _, tool := range c.Tools {
+		// no dependency, pass
+		if len(tool.DependsOn) == 0 {
+			continue
 		}
-	}
-	return absToolFilePath, absVarFilePath, nil
-}
 
-// parseCustomPath  parse the path of tools.yaml or variable.yaml
-func parseCustomPath(configFileName, customPath string) (string, error) {
-	if filepath.IsAbs(customPath) {
-		log.Debugf("Abs path is %s.", customPath)
-		if err := fileExists(customPath); err != nil {
-			log.Errorf("file %s not exists.", customPath)
-			return "", err
+		// for each dependency
+		for _, dependency := range tool.DependsOn {
+			// skip empty string
+			dependency = strings.TrimSpace(dependency)
+			if dependency == "" {
+				continue
+			}
+
+			// generate an error if the dependency isn't in the config set,
+			if _, ok := toolMap[dependency]; !ok {
+				errors = append(errors, fmt.Errorf("tool %s's dependency %s doesn't exist in the config", tool.InstanceID, dependency))
+			}
 		}
-		return customPath, nil
 	}
-	configFilePath, err := filepath.Abs(configFileName)
-	if err != nil {
-		return "", err
-	}
-	absFilePath := filepath.Join(filepath.Dir(configFilePath), customPath)
-	log.Debugf("Abs path is %s.", absFilePath)
-	if err := fileExists(absFilePath); err != nil {
-		log.Errorf("file %s not exists.", absFilePath)
-		return "", err
-	}
-	return absFilePath, nil
+
+	return errors
 }
 
-func fileExists(path string) error {
-	if _, err := os.Stat(path); err != nil {
-		return err
-	}
-	return nil
+// SplitConfigFileBytes take the original config file and split it to:
+// 1. core config
+// 2. variables config
+// 3. tools config
+// Original config should be like below:
+// ---
+// # core config
+// varFile: "" # If not empty, use the specified external variables config file
+// toolFile: "" # If not empty, use the specified external tools config file
+// state:
+//   backend: local
+//   options:
+//     stateFile: devstream.state
+//
+// ---
+// # variables config
+// foo: bar
+//
+// ---
+// # plugins config
+// tools:
+// - name: A-PLUGIN-NAME
+//   instanceID: default
+//   options:
+//     foo: bar
+//
+// See https://github.com/devstream-io/devstream/issues/596 for more details.
+func SplitConfigFileBytes(fileBytes []byte) (coreConfig []byte, varConfig []byte, toolConfig []byte, err error) {
+	splitBytes := bytes.Split(bytes.TrimPrefix(fileBytes, []byte("---")), []byte("---"))
+
+	switch len(splitBytes) {
+	case 1:
+		coreConfig = splitBytes[0]
+	case 2:
+		coreConfig = splitBytes[0]
+		toolConfig = splitBytes[1]
+	case 3:
+		coreConfig = splitBytes[0]
+		varConfig = splitBytes[1]
+		toolConfig = splitBytes[2]
+	default:
+		err = fmt.Errorf("invalid config format")
+	}
+
+	if len(coreConfig) == 0 {
+		err = fmt.Errorf("core config is empty")
+	}
+
+	return
```

Comment:
A nice practice to use `default value`! It's very "Go"!

Reply:
🚀